src/uct/tcp/tcp_cm.c

```diff
@@ -8,86 +8,329 @@
+                                uct_tcp_cm_conn_event_t actual_event)
+{
+    char str_addr[UCS_SOCKADDR_STRING_LEN];
+    ucs_assertv_always(expected_event == actual_event,
+                       "tcp_ep %p: received wrong CM event (expected: %u, "
+                       "actual: %u) from the peer with iface listener "
+                       "address: %s", ep, expected_event, actual_event,
+                       ucs_sockaddr_str(ep->peer_addr->addr, str_addr,
+                                        UCS_SOCKADDR_STRING_LEN));
+    return UCS_OK;
+}
+
+static void uct_tcp_cm_log_somaxconn_exceed(ucs_log_level_t log_level)
+{
+    ucs_log(log_level, "try to increase the accept queue length: "
+            "\"sysctl -w net.core.somaxconn=<new value>\" where "
+            "<new_value> greater than %d", ucs_socket_max_conn());
+}
+
+static ucs_status_t uct_tcp_cm_send_conn_req(uct_tcp_ep_t *ep)
+{
+    uct_tcp_iface_t *iface         = ucs_derived_of(ep->super.super.iface,
+                                                    uct_tcp_iface_t);
+    uct_tcp_cm_conn_pkt_t conn_pkt = {
+        .event                     = UCT_TCP_CM_CONN_REQ
+    };
+    size_t addr_size;
+    ucs_status_t status;
+
+    status = ucs_sockaddr_sizeof((const struct sockaddr*)&iface->config.ifaddr,
+                                 &addr_size);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    memcpy(&conn_pkt.iface_addr, &iface->config.ifaddr, addr_size);
+
+    status = ucs_socket_send(ep->fd, &conn_pkt, sizeof(conn_pkt));
+    if (status != UCS_OK) {
+        uct_tcp_cm_trace_conn_pkt(ep, "unable to send connection request to",
+                                  ep->peer_addr->addr);
+        if (status == UCS_ERR_IO_ERROR) {
+            uct_tcp_cm_log_somaxconn_exceed(UCS_LOG_LEVEL_ERROR);
+        }
+    } else {
+        uct_tcp_cm_trace_conn_pkt(ep, "conection request sent to",
+                                  ep->peer_addr->addr);
+    }
+    return status;
+}
+
+static ucs_status_t uct_tcp_cm_recv_conn_req(uct_tcp_ep_t *ep,
+                                             struct sockaddr *peer_addr)
+{
+    uct_tcp_cm_conn_pkt_t conn_pkt;
+    ucs_status_t status;
+
+    status = ucs_socket_recv(ep->fd, &conn_pkt, sizeof(conn_pkt));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    status = uct_tcp_cm_conn_pkt_check_event(ep, UCT_TCP_CM_CONN_REQ,
+                                             conn_pkt.event);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    *peer_addr = conn_pkt.iface_addr;
+
+    uct_tcp_cm_trace_conn_pkt(ep, "received the connection request from",
+                              peer_addr);
+    return UCS_OK;
+}
+
+static ucs_status_t uct_tcp_cm_send_conn_ack(uct_tcp_ep_t *ep)
+{
+    uct_tcp_cm_conn_pkt_t conn_pkt = {
+        .event                     = UCT_TCP_CM_CONN_ACK
+    };
+    ucs_status_t status;
+
+    status = ucs_socket_send(ep->fd, &conn_pkt, sizeof(conn_pkt));
```

Comment:
minor: when sending the ack, no need to send the sock addr as well (i guess it's uninitialized? how did valgrind not complain?)